test/std/algorithms/alg.sorting/alg.binary.search/binary.search/binary_search.pass.cpp

```diff
@@ -0,0 +1,103 @@
+// <algorithm>
+
+// template<ForwardIterator Iter, class T>
+//   constexpr bool      // constexpr after C++17
+//   binary_search(Iter first, Iter last, const T& value);
+
+#include "oneapi_std_test_config.h"
+#include "test_iterators.h"
+#include "checkData.h"
+
+#include <iostream>
+
+#ifdef USE_ONEAPI_STD
+#    include _ONEAPI_STD_TEST_HEADER(algorithm)
+#    include _ONEAPI_STD_TEST_HEADER(iterator)
+namespace s = oneapi_cpp_ns;
+#else
+#    include <algorithm>
+#    include <iterator>
+namespace s = std;
+#endif
+
+#if TEST_DPCPP_BACKEND_PRESENT
+constexpr sycl::access::mode sycl_read = sycl::access::mode::read;
+constexpr sycl::access::mode sycl_write = sycl::access::mode::write;
+
+template <class Iter, class T>
+bool
+test(Iter first, Iter last, const T& value, bool x)
+{
+    return (s::binary_search(first, last, value) == x);
+}
+
+class KernelBSearchTest1;
+class KernelBSearchTest2;
+class KernelBSearchTest3;
+class KernelBSearchTest4;
+
+template <typename Iter, typename KC>
+void
+kernel_test()
+{
+    sycl::queue deviceQueue = TestUtils::get_test_queue();
+    sycl::cl_bool ret = false;
+    sycl::range<1> numOfItems{1};
+    sycl::buffer<sycl::cl_bool, 1> buffer1(&ret, numOfItems);
+    const unsigned N = 1000;
+    const int M = 10;
+    int host_vbuf[N];
+    int x = 0;
+    for (size_t i = 0; i < N; ++i)
+    {
+        host_vbuf[i] = x;
+        if (++x == M)
+            x = 0;
+    }
+
+    std::sort(host_vbuf, host_vbuf + N);
+    sycl::range<1> host_buffer_sz{N};
+    sycl::buffer<sycl::cl_int, 1> host_data_buffer(host_vbuf, host_buffer_sz);
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        auto ret_access = buffer1.get_access<sycl_write>(cgh);
+        auto data_access = host_data_buffer.get_access<sycl_read>(cgh);
+        cgh.single_task<KC>([=]() {
+            int device_vbuf[N];
```

Comment:
Why do we need to create a private array (in the device registers)? I think we may call test with the device data, which provided by `data_access`.

Reply:
the same.

Reply:
On the one hand, it's more practical that a search is done over the data in global / local memory than in private arrays. On the other hand, I do not know if an iterator obtained by accessor.begin() is convertible to different types.

Let's add an issue/task to address this question later.

Reply:
Fixed